### PR TITLE
[receiver/file] add basic file reading capability 2/n

### PR DESCRIPTION
**Pull request by pmcollins**
**Description:** To keep this change small, it adds a basic capability to the file receiver, where the receiver reads a metrics json file (the output of a file exporter), converts that data into metrics, and sends it to the next consumer. Subsequent PRs will include more features including the ability replay telemetry in time, update timestamps, read from rotated files, and handle different types of telemetry.

**Link to tracking Issue:** #14638

**Testing:** Unit tests added, and manually tested in a simple pipeline.

**Documentation:** Documentation for the current feature set is available in the README. More docs to come as features are added.

### CODE REVIEW

File: receiver/filereceiver/file_reader.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+func (fr fileReader) readLine() error {
+	line, err := fr.stringReader.ReadString('\n')
+	if err != nil {
+		return fmt.Errorf("failed to read line from input file: %w", err)
+	}
+	metrics, err := fr.unm.UnmarshalMetrics([]byte(line))
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal metrics: %w", err)
+	}
+	return fr.consumer.ConsumeMetrics(context.Background(), metrics)
```
Comment (atoulme) [inline receiver/filereceiver/file_reader.go:85]: Can we pass the top level context into this call?

File: receiver/filereceiver/file_reader.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+func (fr fileReader) readAll(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			err := fr.readLine()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					fr.logger.Info("EOF reached")
```
Comment (atoulme) [inline receiver/filereceiver/file_reader.go:64]:
```suggestion
					fr.logger.Debug("EOF reached")
```

**Comment by djaglowski** [comment]
Should we use [`pkg/stanza/fileconsumer`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/stanza/fileconsumer) so that we don't have to reimplement file tracking and reading?

**Comment by pmcollins** [comment]
> Should we use [`pkg/stanza/fileconsumer`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/stanza/fileconsumer) so that we don't have to reimplement file tracking and reading?

Hi @djaglowski, would you be okay with an initial implementation that does not use it, then perhaps a PR by someone who is familiar with the library to replace the file reading code in this receiver with calls into the library? This PR only has 1-2 lines of file reading code so I don't expect there to be much duplicated effort.

**Comment by djaglowski** [comment]
> Hi @djaglowski, would you be okay with an initial implementation that does not use it, then perhaps a PR by someone who is familiar with the library to replace the file reading code in this receiver with calls into the library? This PR only has 1-2 lines of file reading code so I don't expect there to be much duplicated effort.

I think it's fine as long as the receiver's status does not progress to alpha until it's been updated to use the fileconsumer package.

**Review by atoulme — APPROVED** [review]
Changes look good, and failures are unrelated. Please rebase to pick up the fix to main.

**Pull request merged**